check_upgrade/check_upgrade.go

```diff
@@ -0,0 +1,212 @@
+	if err != nil {
+		return []Node{}, errors.Wrapf(err, "while unmarshalling response: %v", string(resp.Json))
+	}
+
+	duplicates := findDuplicateNodes(result.Nodes)
+
+	return duplicates, nil
+}
+
+func printDuplicates(entityType string, ns uint64, nodes []Node) {
+	if len(nodes) > 0 {
+		fmt.Printf("Found duplicate %ss in namespace: #%v\n", entityType, ns)
+		for _, node := range nodes {
+			fmt.Printf("dgraph.xid %v , Uid: %v\n", node.DgraphXID, node.UID)
+		}
+	}
+}
+
+func init() {
+	CheckUpgrade.Cmd = &cobra.Command{
+		Use:   "checkupgrade",
+		Short: "Run the checkupgrade tool",
+		Long:  "The checkupgrade tool is used to check for duplicate dgraph.xid's in the Dgraph database before upgrade.",
+		Run: func(cmd *cobra.Command, args []string) {
+			run()
+		},
+		Annotations: map[string]string{"group": "tool"},
+	}
+	CheckUpgrade.Cmd.SetHelpTemplate(x.NonRootTemplate)
+	flag := CheckUpgrade.Cmd.Flags()
+	flag.String(alphaGrpc, "127.0.0.1:9080",
+		"Dgraph Alpha gRPC server address")
+
+	flag.String(alphaHttp, "http://127.0.0.1:8080", "Draph Alpha HTTP(S) endpoint.")
+}
+
+func duplicateDeleteMutation(nodes []Node) {
+	if len(nodes) > 0 {
+		var deleteMutation strings.Builder
+		deleteMutation.WriteString("{\n")
+		deleteMutation.WriteString(" delete {\n")
+		for _, node := range nodes {
+			deleteMutation.WriteString(fmt.Sprintf("<%v> <%v> \"%v\" .\n", node.UID, "dgraph.xid", node.DgraphXID))
+		}
+		deleteMutation.WriteString(" }\n")
+		deleteMutation.WriteString(" }\n")
+
+		fmt.Printf("\ndelete duplicate nodes using following mutation : \n%v", deleteMutation.String())
+	}
+}
+
+func run() {
+	fmt.Println("Running check-upgrade tool")
+
+	cmdInput := parseInput()
+
+	gc, hc, err := setupClients(cmdInput.alphaGrpc, cmdInput.alphaHttp)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return
+	}
+
+	hc.LoginIntoNamespace(dgraphtest.DefaultUser, dgraphtest.DefaultPassword, x.GalaxyNamespace)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrapf(err, "while logging into namespace: %v", x.GalaxyNamespace))
+		return
+	}
+
+	namespaces, err := hc.ListNamespaces()
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return
+	}
+
+	ctx := context.Background()
+
+	for _, ns := range namespaces {
+		if err := gc.LoginIntoNamespace(ctx, dgraphtest.DefaultUser, dgraphtest.DefaultPassword, ns); err != nil {
+			fmt.Fprintln(os.Stderr, errors.Wrapf(err, "while logging into namespace: %v", ns))
+			return
+		}
+
+		duplicates, err := getDuplicateNodes(ctx, gc)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return
+		}
+
+		var duplicateUsers []Node
+		var duplicateGroups []Node
+
+		for _, node := range duplicates {
+			if slices.Contains(node.DgraphType, "dgraph.type.User") {
+				duplicateUsers = append(duplicateUsers, node)
+			} else if slices.Contains(node.DgraphType, "dgraph.type.Group") {
+				duplicateGroups = append(duplicateGroups, node)
+			}
+		}
+
+		printDuplicates("user", ns, duplicateUsers)
```

Comment:
can you show an example of this?

Reply:
Found duplicate groups in namespace: #1
dgraph.xid user1 , Uid: 0x2711
dgraph.xid user1 , Uid: 0x2712
dgraph.xid user1 , Uid: 0x2713
dgraph.xid user1 , Uid: 0x2714

delete duplicate nodes using following mutation : 
{
 delete {
<0x2711> <dgraph.xid> "user1" .
<0x2712> <dgraph.xid> "user1" .
<0x2713> <dgraph.xid> "user1" .
<0x2714> <dgraph.xid> "user1" .
 }
 }
Found duplicate users in namespace: #0
dgraph.xid user1 , Uid: 0x3
dgraph.xid user1 , Uid: 0x4

delete duplicate nodes using following mutation : 
{
 delete {
<0x3> <dgraph.xid> "user1" .
<0x4> <dgraph.xid> "user1" .
 }
 }

Reply:
also added this into the code

Reply:
We need to print this better. We need to show what users are duplicates, show both of their UIDs. We also need to break it into three parts, first part showing duplicate users, second part showing duplicate groups and third part showing user and groups that have same name.

Additionally, we also need to delete everything about that UID, one of which is duplicate. We cannot pick one at random. We probably just show an example of how to delete the UID I guess.